cpp/src/arrow/dataset/scanner.cc

```diff
@@ -151,10 +151,56 @@ Status NormalizeScanOptions(const std::shared_ptr<ScanOptions>& scan_options,
     // If we couldn't infer it from the projection expression then just grab all
     // fields from the dataset
     if (!scan_options->projected_schema) {
-      ARROW_ASSIGN_OR_RAISE(auto projection_descr,
-                            ProjectionDescr::Default(*dataset_schema));
-      scan_options->projected_schema = std::move(projection_descr.schema);
-      scan_options->projection = projection_descr.expression;
+      // Until now, we assume the project expression is bound, but if it is not
+      // bound, we have to check the expressions and make sure bind them
+      // and create the projected schema based on the field_refs (which guarantees
+      // IsName() to be true).
+
+      // process resultant dataset_schema after projection
+      std::shared_ptr<Schema> projected_schema;
```

Comment:
There is a lot of duplication with the path above (e.g. when the schema is bound).  I wonder if there is some way to simplify the two paths.  Right now it looks like:

"If we have a bound expression we use the types and names from the expression nodes to form the schema"

and

"If we have an unbound expression we use the names from the expression to find fields in the dataset schema and get the types from there."

Perhaps the second approach would work in both cases (e.g. we could grab fields from the dataset schema even when the expression is bound)?

Reply:
Yes, that was possible at least in the test cases in `C++`, `R` and `Python` on my machine. I have updated the code. Let's see how it goes in the CIs.